libraries/anvill_passes/src/BaseFunctionPass.cpp

```diff
@@ -21,49 +21,16 @@
 
 namespace anvill {
 
-char BaseFunctionPass::ID = '\0';
-
-bool BaseFunctionPass::OperandReferencesStackPointer(const llvm::Value *value) {
-  if (IsStackPointer(const_cast<llvm::Value *>(value))) {
-    return true;
-  }
-
-  std::vector<const llvm::Value *> operand_list;
-
-  if (auto instr = llvm::dyn_cast<llvm::Instruction>(value); instr != nullptr) {
-
-    auto operand_count = instr->getNumOperands();
-    for (auto i = 0U; i < operand_count; ++i) {
-      operand_list.push_back(instr->getOperand(i));
-    }
-
-  } else if (auto constant_expr = llvm::dyn_cast<llvm::ConstantExpr>(value);
-             constant_expr != nullptr) {
-
-    auto operand_count = constant_expr->getNumOperands();
-    for (auto i = 0U; i < operand_count; ++i) {
-      operand_list.push_back(constant_expr->getOperand(i));
-    }
-  }
-
-  for (const auto &operand : operand_list) {
-    if (OperandReferencesStackPointer(operand)) {
-      return true;
-    }
-  }
-
-  return false;
-}
-
 bool BaseFunctionPass::InstructionReferencesStackPointer(
```

Comment:
Update a cache mapping llvm::Value -> state, and introduce base cases to handle PHI nodes.

Reply:
Key thing might be "when to clear the cache." You could hide the `runOnFunction` of an llvm pass as a derived virtual method, e.g. `virtual bool Run(llvm::Function &func) = 0;`, then your `BaseFunctionPass` implements a `runOnFunction` that clears out the cache, then returns `this->Run(func);`.

Reply:
Then `BaseFunctionPass` would derive from `llvm::FunctionPass`.